retis/src/helpers/bimap.rs

```diff
@@ -0,0 +1,148 @@
+use std::{
+    borrow::Borrow,
+    collections::BTreeMap,
+    ops::{Bound, RangeBounds},
+    sync::Arc,
+};
+
+/// An iterator over a range of key->values pairs in a `BiBTreeMap`.
+pub(crate) struct KeyRange<'a, K: 'a, V: 'a> {
+    inner: std::collections::btree_map::Range<'a, Arc<K>, Arc<V>>,
+}
+
+impl<'a, K: Ord, V> KeyRange<'a, K, V> {
+    fn new(map: &'a BTreeMap<Arc<K>, Arc<V>>, range: (Bound<&'a K>, Bound<&'a K>)) -> Self {
+        Self {
+            inner: map.range::<K, (Bound<&K>, Bound<&K>)>(range),
+        }
+    }
+}
+
+impl<'a, K: Ord, V> Iterator for KeyRange<'a, K, V> {
+    type Item = (&'a K, &'a V);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next().map(|(k, v)| (k.as_ref(), v.as_ref()))
+    }
+}
+
+impl<'a, K: Ord, V> DoubleEndedIterator for KeyRange<'a, K, V> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        self.inner
+            .next_back()
+            .map(|(k, v)| (k.as_ref(), v.as_ref()))
+    }
+}
+
+/// A bimap backed by two `BTreeMap`.
+pub(crate) struct BiBTreeMap<K, V> {
+    /// A `BTreeMap` of `Arc` wrapped keys and values representing the key->value
+    /// relationship.
+    left2right: BTreeMap<Arc<K>, Arc<V>>,
+    /// A `BTreeMap` of `Arc` wrapped keys and values representing the value->key
+    /// relationship.
+    right2left: BTreeMap<Arc<V>, Arc<K>>,
+}
+
+impl<K, V> BiBTreeMap<K, V>
+where
+    K: std::hash::Hash + Eq + Ord + std::fmt::Display,
```

Comment:
Nit: no need to require `std::fmt::Display` here.

Reply:
Ah right. I added them for some manual testing while coding. Sorry it is a leftover. Dropping it.